library/md.c

```diff
@@ -773,9 +773,6 @@ mbedtls_md_type_t mbedtls_md_get_type(const mbedtls_md_info_t *md_info)
 #if defined(MBEDTLS_PSA_CRYPTO_C)
 psa_algorithm_t mbedtls_md_psa_alg_from_type(mbedtls_md_type_t md_type)
 {
-    if (md_type == MBEDTLS_MD_NONE) {
```

Comment:
May I ask why we remove this part of code and add a comment `Must not be NONE.`?

Reply:
Sure. This is to allow simplifying the conversion later by using bit operations, that may not guarantee that NONE will be translated to NONE. I first did this part in a separate commit as it changes the function's contract and I wanted to check the impact of that before making further changes.

Reply:
I agree that's a possibility, but (1) it appears we don't actually need NONE->NONE conversion in this internal function, and (2) the primary purpose of this PR is to save on code size. As the table in the PR description shows, most of the saving come from making the conversion functions `static inline`, but that will only work if they are super short. Just adding an extra `if` statement would negate a significant part of the code size savings.

PS: the change to using bit operations is done as part of this PR, and actually now that I look again at the history, it was done before removing the NONE->NONE guarantee, first keeping the `if` as you suggested (that's [this commit](https://github.com/Mbed-TLS/mbedtls/pull/7546/commits/1f6d2e352dd7ebfbef16249a0425271028deed5a)) but this did not bring code size savings until the functions were made `static inline` as the table with code size measurements show.

Reply:
I'd be worried if `mbedtls_md_psa_alg_from_type(MBEDTLS_MD_NONE)` returned a valid PSA algorithm (either a hash or not), but it doesn't: it returns a value that is deliberately unassigned, and that is effectively equivalent to 0 in most ways a PSA hash algorithm can be used. Essentially `PSA_ALG_CATEGORY_HASH` is “none, but you know it's a hash”. So that's fine.

Reply:
I've made some measurement keeping the `if` as suggested with `static inline` functions and it negates all code size wins, see "(inline + NONE->NONE)" in the table above. I've also tried keeping the `if` in `mbedtls_md_psa_alg_from_type()` while keeping it a normal function an making only `mbedtls_md_psa_alg_from_type` static inline, see line "(only one inline)" in the table, which unfortunately doesn't bring any code size improvement either.

So, I'm afraid we have to choose: either we want code size wins and we don't have NONE->NONE conversion, or we think NONE->NONE is important, and we scratch this PR entirely as it was a dead-end.